did-ion/src/sidetree.rs

```diff
@@ -0,0 +1,1926 @@
+                    "Unexpected service properties"
+                );
+                let service_endpoint = match service_endpoint {
+                    None => bail!("Missing endpoint for service"),
+                    Some(OneOrMany::Many(_)) => bail!("Sidetree service must contain one endpoint"),
+                    Some(OneOrMany::One(se)) => se,
+                };
+                let sub_id = did_url_to_id(&id, did).context("Convert service id")?;
+                let service_type = match type_ {
+                    OneOrMany::One(type_) => type_,
+                    OneOrMany::Many(_) => bail!("Service must contain single type"),
+                };
+                ensure!(b64len(&service_type) <= 30, "Sidetree service type must contain no more than 30 Base64Url-encoded characters");
+                ensure!(
+                    b64len(&sub_id) <= 50,
+                    "Sidetree service id must contain no more than 50 Base64Url-encoded characters"
+                );
+                let entry = ServiceEndpointEntry {
+                    id: sub_id,
+                    r#type: service_type,
+                    service_endpoint,
+                };
+                DIDStatePatch::AddServices {
+                    services: vec![entry],
+                }
+            }
+            DIDDocumentOperation::RemoveVerificationMethod(did_url) => {
+                let id = did_url.to_string();
+                DIDStatePatch::RemovePublicKeys { ids: vec![id] }
+            }
+            DIDDocumentOperation::RemoveService(did_url) => {
+                let id = did_url.to_string();
+                DIDStatePatch::RemoveServices { ids: vec![id] }
+            }
+        })
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct SidetreeError {
+    // List of error codes: https://github.com/decentralized-identity/sidetree/blob/v1.0.0/lib/core/versions/1.0/ErrorCode.ts
+    pub code: String,
+    pub message: Option<String>,
+}
+
+impl fmt::Display for SidetreeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Sidetree error {}", self.code)?;
+        if let Some(ref message) = self.message {
+            write!(f, ": {}", message)?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl<S: Sidetree + Send + Sync> DIDMethod for SidetreeClient<S> {
+    fn name(&self) -> &'static str {
+        S::METHOD
+    }
+
+    fn to_resolver(&self) -> &dyn DIDResolver {
+        match self.resolver {
+            Some(ref res) => res,
+            None => &NoOpResolver,
+        }
+    }
+
+    fn create(&self, create: DIDCreate) -> Result<DIDMethodTransaction> {
+        let DIDCreate {
+            recovery_key,
+            update_key,
+            verification_key,
+            options,
+        } = create;
+        ensure!(is_empty(&options), "Create options not supported");
```

Comment:
Right now the options are method-specific, and there aren't other options to use, so I don't know exactly.
Erroring on unknown option seems to me like erring on the side of caution.
But in DID Registrar driver implementations that I looked at, there is no such requirement that all passed options are recognized. So it might be more consistent to drop this `ensure` - or indeed make it a warning (changing the function's return type and plumbing the warnings through to DIDKit).

It could be nice if there was some convention to indicate whether an option may be safely ignored or is required to be understood - like using the ["crit"](https://www.rfc-editor.org/rfc/rfc7515.html#section-4.1.11) JWS header parameter.

Unlike with [DID resolution input metadata](https://www.w3.org/TR/did-spec-registries/#did-resolution-input-metadata) / [options](https://w3c.github.io/did-spec-registries/#did-resolution-options), there isn't yet a registry of DID operation [options](https://identity.foundation/did-registration/#options). DID resolution options have a similar issue I think, of how to know whether an option must be understood or can be ignored. I think the tendency is to treat options as optional to implement by default. Then the caller might have to infer from the output whether the option was actually understood and processed. (vs. inferring from an error message that a given option was unsupported, which would happen with this current implementation). The current Sidetree/ION DID resolver HTTP(S) endpoint though actually fail if any query string parameters are passed, making any resolution input-metadata/options result in an error... Some of those options, like `service`, could be handled client-side (in ssi's dereference function).
For DID method operations besides read/resolve, one approach could be that a critical option (option that must be understood) should just have a new operation defined for it specifically. This PR doesn't add a generic mechanism for additional DID method operations, although of course DID method implementations could add their own method-specific operation implementations.